include/event2/event_watcher.h

```diff
@@ -29,98 +26,63 @@
-struct prepare_watcher_cb_info {
-	/**
-		The current time (may be cached). This value is scoped to this "prepare" callback; if you need
-		to retain it outside the scope of the callback (e.g. to measure the duration between callbacks),
-		make a copy.
-	 */
-	struct timeval now;
-
-	/**
-		The timeout duration passed to the underlying implementation's `dispatch`. This will be NULL if
-		there are no pending EV_TIMEOUT events. It may be a useful performance statistic to compare this
-		value -- essentially the expected polling duration -- against the actual polling duration
-		measured by comparing the "now" times reported in prepare and check.
-	 */
-	const struct timeval *timeout;
-};
-
-/**
-	Contextual information passed from event_base_loop to the "check" watcher callbacks. We define
-	this as a struct rather than individual parameters to the callback function for the sake of future
-	extensibility.
- */
-struct check_watcher_cb_info {
-	/**
-		The current time (may be cached). This value is scoped to this "check" callback; if you need to
-		retain it outside the scope of the callback (e.g. to measure the duration between callbacks),
-		make a copy.
-	 */
-	struct timeval now;
-};
-
-/**
-	Prepare callback, invoked by event_base_loop immediately before polling for I/O.
+	Prepare callback, invoked by event_base_loop immediately before polling
+	for I/O.
 	@param watcher the prepare watcher that invoked this callback.
 	@param info contextual information passed from event_base_loop.
  */
-typedef void (*prepare_watcher_cb)(struct event_watcher *, const struct prepare_watcher_cb_info *);
+typedef void (*event_prepare_cb)(struct event_watcher *, const struct event_prepare_cb_info *);
 
 /**
-	Check callback, invoked by event_base_loop immediately after polling for I/O and before processing
-	any active events.
+	Check callback, invoked by event_base_loop immediately after polling for
+	I/O and before processing any active events.
 	@param watcher the check watcher that invoked this callback.
 	@param info contextual information passed from event_base_loop.
  */
-typedef void (*check_watcher_cb)(struct event_watcher *, const struct check_watcher_cb_info *);
+typedef void (*event_check_cb)(struct event_watcher *, const struct event_check_cb_info *);
 
 /**
-	Register a new "prepare" watcher, to be called in the event loop prior to polling for events.
-	Watchers will be called in the order they were registered.
+	Register a new "prepare" watcher, to be called in the event loop prior
+	to polling for events. Watchers will be called in the order they were
+	registered.
 	@param base the event_base to operate on.
 	@param callback the callback function to invoke.
 	@return a pointer to the newly allocated event watcher.
  */
 EVENT2_EXPORT_SYMBOL
-struct event_watcher *event_watcher_prepare_new(struct event_base *base, prepare_watcher_cb callback);
+struct event_watcher *event_watcher_prepare_new(struct event_base *base, event_prepare_cb callback);
```

Comment:
looking into this again, and priorities support for this feature came to my mind, i.e. if we want this be generic, since if some library has event_base and registered watches and user want to call something before that library watches he cannot do this right now.

although this is pretty specific thing, and to avoid delaying merging this, we can at least reserve argument in the API for this (to make further changes ABI compatible) and return an error if `prio != 0`

Reply:
Let me suggest an alternative approach: if we want to support priority for watchers later, we could add a function `event_watcher_priority_set`, similar to the existing `event_priority_set`. That would keep the `event_watcher_*_new` APIs minimal, and allow you to change the priority of a watcher...

Reply:
Yep, sounds great